### ISSUE TITLE
Normalize Markdown in `.pot` files

### ISSUE THREAD

**Issue by mgeisler** [google/comprehensive-rust#494]
When `mdbook-xgettext` extracts translatable text, it would be great if it could normalize the strings. This would make it possible for us to reformat the entire course without fearing that the translations get destroyed while doing so.

The normalization would take Markdown like this

~~~markdown
# This is a heading

This is another heading
=======================

A _little_
paragraph.

```rust,editable
fn main() {
    println!("Hello world!");
}
```

* First
* Second
~~~

and turn it into these messages in the `.pot` file:

* `"This is a heading"` ([atx heading](https://spec.commonmark.org/0.30/#atx-headings) is stripped)
* `"This is another heading"` ([setext heading](https://spec.commonmark.org/0.30/#setext-headings) is stripped)
* `"A _little_ paragraph."` (soft-wrapped lines are unfolded)
* `"fn main() {\n    println!("Hello world!");\n}"` (info string is stripped, we should instead use a [`#,` flag](https://www.gnu.org/software/gettext/manual/html_node/PO-Files.html))
* `"First"` (bullet point extracted individually)
* `"Second"`

Like in #318, we should do this in a step-by-step fashion and make sure to apply the transformations to the existing translations. It would also be good if we have a way to let translators update their not-yet-submitted translations.

**Comment by mgeisler** [comment]
When writing this, it occured to me that we can get to the end-state of nicely formatted Markdown by running a formatter on 

- The `.md` files
- The `msgid` fields in all `.po` files
- The `msgstr` fields in all `.po` files

If we do this in lock-step, we ought to end up with a lossless process. I was experimenting with this over the weekend: the `msgfilter` program can process the `msgstr` fields, but we need our own little helper to do the same with the `msgid` fields.

I was using https://dprint.dev/ for the formatting since it seems fast and extensible. It can format [code blocks inside Markdown](https://dprint.dev/plugins/markdown/) which is something I really want :smile:

**Comment by jooyunghan** [comment]
Hey Martin,
While updating Korean last week, i found a few missing statements in msgstr entries. Is it from any recent updates of xgettext?

**Comment by mgeisler** [comment]
> While updating Korean last week, i found a few missing statements in msgstr entries. Is it from any recent updates of xgettext?

A few entries were lost and changed as part of #449 — @djmitche would know the details :smile:

**Comment by djmitche** [comment]
Yeah, that was a somewhat lossy process. Hopefully those appear in fuzzy or old messages?

**Comment by djmitche** [comment]
@mgeisler I'm not sure what you mean about the formatter. Can you describe that in more detail?

**Comment by mgeisler** [comment]
> Can you describe that in more detail?

Sure! My thinking is that we can safely format the Markdown files if we we know that it won't create more/fewer entries in the PO files when `mdbook-xgettext` is executed on the formatted files.

To avoid losing translations, we'll then run the same formatting on all `msgid` fields.

We can already safely run the formatter on the `msgstr` entries today with `msgfilter` and that might be something we should encourage translators to do to get smaller diffs.
